Mathlib/MeasureTheory/Constructions/Projective.lean

```diff
@@ -0,0 +1,146 @@
+    P I S = P J T := by
+  have : S = (fun f : ∀ i : I, α i ↦ fun j : J ↦ f ⟨j, hJI j.prop⟩) ⁻¹' T :=
+    eq_of_cylinder_eq_of_subset h_eq hJI
+  rw [hP I J hJI, Measure.map_apply _ hT, this]
+  exact measurable_pi_lambda _ (fun _ ↦ measurable_pi_apply _)
+
+lemma congr_cylinder [Nonempty (∀ i, α i)] (hP : IsProjectiveMeasureFamily P)
+    {S : Set (∀ i : I, α i)} {T : Set (∀ i : J, α i)} (hS : MeasurableSet S) (hT : MeasurableSet T)
+    (h_eq : cylinder I S = cylinder J T) :
+    P I S = P J T := by
+  classical
+  let U := (fun f : ∀ i : (I ∪ J : Finset ι), α i
+        ↦ fun j : I ↦ f ⟨j, Finset.mem_union_left J j.prop⟩) ⁻¹' S ∩
+      (fun f ↦ fun j : J ↦ f ⟨j, Finset.mem_union_right I j.prop⟩) ⁻¹' T
+  suffices : P (I ∪ J) U = P I S ∧ P (I ∪ J) U = P J T
+  exact this.1.symm.trans this.2
+  constructor
+  · have h_eq_union : cylinder I S = cylinder (I ∪ J) U := by
+      rw [← inter_cylinder, h_eq, inter_self]
+    exact hP.congr_cylinder_of_subset hS h_eq_union.symm (Finset.subset_union_left _ _)
+  · have h_eq_union : cylinder J T = cylinder (I ∪ J) U := by
+      rw [← inter_cylinder, h_eq, inter_self]
+    exact hP.congr_cylinder_of_subset hT h_eq_union.symm (Finset.subset_union_right _ _)
+
+/-- Auxiliary lemma for `measure_univ_eq`. -/
+lemma measure_univ_eq_of_subset (hP : IsProjectiveMeasureFamily P) (hJI : J ⊆ I) :
+    P I univ = P J univ := by
+  classical
+  have : (univ : Set (∀ i : I, α i)) =
+      (fun x : ∀ i : I, α i ↦ fun i : J ↦ x ⟨i, hJI i.2⟩) ⁻¹' (univ : Set (∀ i : J, α i)) :=
+    by rw [preimage_univ]
+  rw [this, ← Measure.map_apply _ MeasurableSet.univ]
+  · rw [hP I J hJI]
+  · exact measurable_pi_lambda _ (fun _ ↦ measurable_pi_apply _)
+
+lemma measure_univ_eq (hP : IsProjectiveMeasureFamily P) (I J : Finset ι) :
+    P I univ = P J univ := by
+  classical
+  rw [← hP.measure_univ_eq_of_subset (Finset.subset_union_left I J),
+    ← hP.measure_univ_eq_of_subset (Finset.subset_union_right I J)]
+
+end IsProjectiveMeasureFamily
+
+/-- A measure `μ` is the projective limit of a family of measures indexed by finite sets of `ι` if
+for all `I : Finset ι`, the projection from `∀ i, α i` to `∀ i : I, α i` maps `μ` to `P I`. -/
+def IsProjectiveLimit (μ : Measure (∀ i, α i))
+    (P : ∀ J : Finset ι, Measure (∀ j : J, α j)) : Prop :=
+  ∀ I : Finset ι, (μ.map fun x : ∀ i, α i ↦ fun i : I ↦ x i) = P I
+
+namespace IsProjectiveLimit
+
+variable {μ ν : Measure (∀ i, α i)}
+
+lemma measure_cylinder (h : IsProjectiveLimit μ P)
+    (I : Finset ι) {s : Set (∀ i : I, α i)} (hs : MeasurableSet s) :
+    μ (cylinder I s) = P I s := by
+  rw [cylinder, ← Measure.map_apply _ hs, h I]
+  exact measurable_pi_lambda _ (fun _ ↦ measurable_pi_apply _)
+
+lemma measure_univ_eq (hμ : IsProjectiveLimit μ P) (I : Finset ι) :
+    μ univ = P I univ := by
+  rw [← cylinder_univ I, hμ.measure_cylinder _ MeasurableSet.univ]
+
+lemma isFiniteMeasure [hι : Nonempty ι] [∀ i, IsFiniteMeasure (P i)]
```

Comment:
No need for nonemptyness here: you can use the empty finset instead of a singleton. Same thing for all other statements in this paragraph.